### PR TITLE
Add street heuristic

### PR DESCRIPTION
**Pull request by AdamWiktor**

### CODE REVIEW

File: src/main/java/org/opentripplanner/graph_builder/GraphBuilder.java [hunk 1]
```diff
@@ -376,7 +395,7 @@ private static void tryAddVehicleSharingBuilderModule(GraphBuilder graphBuilder)
      * types are present. This helps point out when config files have been misnamed (builder-config vs. build-config).
      */
     private static enum InputFileType {
-        GTFS, OSM, DEM, CONFIG, GRAPH, JSON, TRANSIT_TICKETS, OTHER;
+        GTFS, OSM, DEM, CONFIG, GRAPH, JSON, TRANSIT_TICKETS, OTHER, SERIALIZED_STREETS;
```
Comment (Kamyki) [inline src/main/java/org/opentripplanner/graph_builder/GraphBuilder.java]: Change name to `SERIALIZED_CONNECTION_MATRIX`

File: src/main/java/org/opentripplanner/graph_builder/module/street_heuristic/StreetHeuristicGraphBuilderModule.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashMap;
+
+@AllArgsConstructor
+public class StreetHeuristicGraphBuilderModule implements GraphBuilderModule {
+
+    private static final Logger LOG = LoggerFactory.getLogger(StreetHeuristicGraphBuilderModule.class);
+
+    private final ObjectMapper objectMapper = new ObjectMapper();
```
Comment (Kamyki) [inline src/main/java/org/opentripplanner/graph_builder/module/street_heuristic/StreetHeuristicGraphBuilderModule.java]: Do we want to use global object mapper?

File: src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/Boundaries.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
+package org.opentripplanner.routing.algorithm.strategies.street_heuristic;
```
Comment (Kamyki) [inline src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/Boundaries.java]: Remember about changing dir name

File: src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/Boundaries.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        this.lonMin = lonMin;
+        this.lonMax = lonMax;
+        this.width = width;
+        this.height = height;
+        cellHeight = (latMax - latMin) / height;
+        cellWidth = (lonMax - lonMin) / width;
+        latBegin = latMin - cellHeight / 2;
+        lonBegin = lonMin - cellWidth / 2;
+        latEnd = latMax + cellHeight / 2;
+        lonEnd = lonMax + cellWidth / 2;
```
Comment (Kamyki) [inline src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/Boundaries.java]:
```suggestion
        latEnd = latMax - cellHeight / 2;
        lonEnd = lonMax - cellWidth / 2;
```
Check if max is really pointing at point + 0.5 of width/height of cell after border

File: src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/Point.java [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package org.opentripplanner.routing.algorithm.strategies.street_heuristic;
+
+import lombok.Value;
+
+@Value
+class Point {
+
+    int x, y;
```
Comment (Kamyki) [inline src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/Point.java]:
```suggestion
    int i, j;
```

File: src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/Boundaries.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    Point createPointFrom(double lat, double lon) {
+        int x = (int) ((lat - latBegin) / cellHeight);
+        int y = (int) ((lon - lonBegin) / cellWidth);
+        return new Point(x, y);
+    }
+
+    boolean contains(Point point) {
+        return point.getX() >= latBegin && point.getX() <= latEnd && point.getY() >= lonBegin && point.getY() <= lonEnd;
```
Comment (Kamyki) [inline src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/Boundaries.java]: compare with `height,width` instead

File: src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/StreetHeuristicData.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        boundaries = Boundaries.from(data);
+        int width = boundaries.getWidth();
+        int height = boundaries.getHeight();
+        this.data.values().forEach(array -> {
+            for (int i = 0; i < height; i++)
+                for (int j = 0; j < width; j++)
+                    if (array[i][j] == 0.)
+                        array[i][j] = Float.NaN;
+        });
+    }
```
Comment (Kamyki) [inline src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/StreetHeuristicData.java]: triple for loop over map elements

File: src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/StreetHeuristicData.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+    Stream<PointWithDirection> getNeighbors(Point point) {
+        return Arrays.stream(Direction.values())
+                .map(direction -> new PointWithDirection(direction.neighbor(point), direction))
+                .filter(p -> p.getPoint().getX() >= 0 && p.getPoint().getX() < boundaries.getHeight())
+                .filter(p -> p.getPoint().getY() >= 0 && p.getPoint().getY() < boundaries.getWidth());
```
Comment (Kamyki) [inline src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/StreetHeuristicData.java]: use `boundaries.contains(p)`

File: src/main/java/org/opentripplanner/routing/graph/Graph.java [hunk 9]
```diff
@@ -320,6 +354,8 @@
      */
     public RouterHealth routerHealth = new RouterHealth();
 
+    public StreetHeuristicData streetHeuristicData;
```
Comment (Kamyki) [inline src/main/java/org/opentripplanner/routing/graph/Graph.java]: ?

File: src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/StreetRWH.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    @Override
+    public double estimateRemainingWeight(State s) {
+        Arrays.stream(visited).forEach(array -> Arrays.fill(array, false));
+
+        PriorityQueue<PointWithWeight> queue = new PriorityQueue<>();
+        Point source = data.mapToPoint(s.getVertex());
+        if (!data.getBoundaries().contains(source)) {
+            return 0.; // TODO euclidean
+        }
+        queue.add(new PointWithWeight(data.mapToPoint(s.getVertex()), INITIAL_WEIGHT));
```
Comment (Kamyki) [inline src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/StreetRWH.java]: add TODO to set INITIAL_WEIGHT based on data

File: src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/StreetRWH.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+        this.data = options.rctx.graph.streetHeuristicData;
+        visited = new boolean[data.getBoundaries().getHeight()][data.getBoundaries().getWidth()];
+        destination = data.mapToPoint(options.to); // TODO outside boundaries
+    }
+
+    @Override
+    public double estimateRemainingWeight(State s) {
+        Arrays.stream(visited).forEach(array -> Arrays.fill(array, false));
+
+        PriorityQueue<PointWithWeight> queue = new PriorityQueue<>();
```
Comment (Kamyki) [inline src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/StreetRWH.java]: priority queue and visited matrix should be both class variable or method variable

**Review by Kamyki — CHANGES_REQUESTED** [review]
Next step - A*

File: src/main/java/org/opentripplanner/graph_builder/GraphBuilder.java (hunk 1 above)
Comment (AdamWiktor) [reply src/main/java/org/opentripplanner/graph_builder/GraphBuilder.java]: done

File: src/main/java/org/opentripplanner/graph_builder/module/street_heuristic/StreetHeuristicGraphBuilderModule.java (hunk 2 above)
Comment (AdamWiktor) [reply src/main/java/org/opentripplanner/graph_builder/module/street_heuristic/StreetHeuristicGraphBuilderModule.java]: we use object mappers like that in whole project

File: src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/Boundaries.java (hunk 3 above)
Comment (AdamWiktor) [reply src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/Boundaries.java]: done

File: src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/Boundaries.java (hunk 4 above)
Comment (AdamWiktor) [reply src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/Boundaries.java]: it was my mistake - lonMin and lonMax are already proper borders

File: src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/Point.java (hunk 5 above)
Comment (AdamWiktor) [reply src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/Point.java]: done

File: src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/Boundaries.java (hunk 6 above)
Comment (AdamWiktor) [reply src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/Boundaries.java]: done

File: src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/StreetHeuristicData.java (hunk 7 above)
Comment (AdamWiktor) [reply src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/StreetHeuristicData.java]: done

File: src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/StreetHeuristicData.java (hunk 8 above)
Comment (AdamWiktor) [reply src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/StreetHeuristicData.java]: done

File: src/main/java/org/opentripplanner/routing/graph/Graph.java (hunk 9 above)
Comment (AdamWiktor) [reply src/main/java/org/opentripplanner/routing/graph/Graph.java]: it is set in graph builder module and used during heuristic initialization

File: src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/StreetRWH.java (hunk 11 above)
Comment (AdamWiktor) [reply src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/StreetRWH.java]: done

File: src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/StreetRWH.java (hunk 10 above)
Comment (AdamWiktor) [reply src/main/java/org/opentripplanner/routing/algorithm/strategies/street_heuristic/StreetRWH.java]: done

File: src/main/java/org/opentripplanner/routing/algorithm/strategies/connection_matrix_heuristic/Boundaries.java [hunk 12]
```diff
@@ -23,24 +23,20 @@
...
         return new Boundaries(data.getLatMin(), data.getLatMax(), data.getLonMin(), data.getLonMax(), data.getWidth(),
                 data.getHeight());
     }
 
     Point createPointFrom(double lat, double lon) {
-        int x = (int) ((lat - latBegin) / cellHeight);
-        int y = (int) ((lon - lonBegin) / cellWidth);
+        int x = (int) ((lat - latMin) / cellHeight);
+        int y = (int) ((lon - lonMin) / cellWidth);
         return new Point(x, y);
```
Comment (Kamyki) [inline src/main/java/org/opentripplanner/routing/algorithm/strategies/connection_matrix_heuristic/Boundaries.java]: Rename local `x,y` to `i,j`

File: src/main/java/org/opentripplanner/routing/algorithm/strategies/connection_matrix_heuristic/Boundaries.java [hunk 13]
```diff
@@ -23,24 +23,20 @@
...
     }
 
     boolean contains(Point point) {
-        return point.getX() >= latBegin && point.getX() <= latEnd && point.getY() >= lonBegin && point.getY() <= lonEnd;
+        return point.getI() >= 0 && point.getI() < height && point.getJ() >= 0 && point.getJ() < width;
```
Comment (Kamyki) [inline src/main/java/org/opentripplanner/routing/algorithm/strategies/connection_matrix_heuristic/Boundaries.java]: Usually `i` indicates horizontal axis, i. e. width.
E.g. i,j = 1,2 means "1 to the right, 2 up"

File: src/main/java/org/opentripplanner/graph_builder/GraphBuilder.java [hunk 14]
```diff
@@ -395,7 +395,7 @@ private static void tryAddVehicleSharingBuilderModule(GraphBuilder graphBuilder)
      * types are present. This helps point out when config files have been misnamed (builder-config vs. build-config).
      */
     private static enum InputFileType {
-        GTFS, OSM, DEM, CONFIG, GRAPH, JSON, TRANSIT_TICKETS, OTHER, SERIALIZED_STREETS;
+        GTFS, OSM, DEM, CONFIG, GRAPH, JSON, TRANSIT_TICKETS, OTHER, SERIALIZER_CONNECTION_MATRIX;
```
Comment (Kamyki) [inline src/main/java/org/opentripplanner/graph_builder/GraphBuilder.java]:
```suggestion
        GTFS, OSM, DEM, CONFIG, GRAPH, JSON, TRANSIT_TICKETS, OTHER, SERIALIZED_CONNECTION_MATRIX;
```

File: src/main/java/org/opentripplanner/routing/algorithm/strategies/connection_matrix_heuristic/ConnectionMatrixHeuristicData.java [hunk 15]
```diff
@@ -14,33 +14,32 @@
...
+                ConnectionMatrixHeuristicDirectionData::getDataAsArray));
         boundaries = Boundaries.from(data);
         int width = boundaries.getWidth();
         int height = boundaries.getHeight();
-        this.data.values().forEach(array -> {
+        for (Float[][] array : this.data.values()) {
             for (int i = 0; i < height; i++)
                 for (int j = 0; j < width; j++)
                     if (array[i][j] == 0.)
                         array[i][j] = Float.NaN;
```
Comment (Kamyki) [inline src/main/java/org/opentripplanner/routing/algorithm/strategies/connection_matrix_heuristic/ConnectionMatrixHeuristicData.java]:
```suggestion
        for (Float[][] array : this.data.values()) {
            for (int j = 0; i < height; i++)
                for (int i = 0; j < width; j++)
                    if (array[j][i] == 0.)
                        array[j][i] = Float.NaN;
```

File: src/main/java/org/opentripplanner/routing/algorithm/strategies/connection_matrix_heuristic/Boundaries.java (hunk 13 above)
Comment (Kamyki) [reply src/main/java/org/opentripplanner/routing/algorithm/strategies/connection_matrix_heuristic/Boundaries.java]: As discussed, we will use descriptive `lonIndex`

**Review by Kamyki — APPROVED** [review]

File: src/main/java/org/opentripplanner/graph_builder/GraphBuilder.java (hunk 14 above)
Comment (AdamWiktor) [reply src/main/java/org/opentripplanner/graph_builder/GraphBuilder.java]: done

**Pull request merged**